<!DOCTYPE html>
        <html>
        <head>
            <meta charset="UTF-8">
            <title>Identifying Spatially Variable Genes</title>
            <style>
/* From extension vscode.github */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

.vscode-dark img[src$=\#gh-light-mode-only],
.vscode-light img[src$=\#gh-dark-mode-only],
.vscode-high-contrast:not(.vscode-high-contrast-light) img[src$=\#gh-light-mode-only],
.vscode-high-contrast-light img[src$=\#gh-dark-mode-only] {
	display: none;
}

</style>
            <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex/dist/katex.min.css">
<link href="https://cdn.jsdelivr.net/npm/katex-copytex@latest/dist/katex-copytex.min.css" rel="stylesheet" type="text/css">
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/markdown.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/highlight.css">
<style>
            body {
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe WPC', 'Segoe UI', system-ui, 'Ubuntu', 'Droid Sans', sans-serif;
                font-size: 14px;
                line-height: 1.6;
            }
        </style>
        <style>
.task-list-item {
    list-style-type: none;
}

.task-list-item-checkbox {
    margin-left: -20px;
    vertical-align: middle;
    pointer-events: none;
}
</style>
<style>
:root {
  --color-note: #0969da;
  --color-tip: #1a7f37;
  --color-warning: #9a6700;
  --color-severe: #bc4c00;
  --color-caution: #d1242f;
  --color-important: #8250df;
}

</style>
<style>
@media (prefers-color-scheme: dark) {
  :root {
    --color-note: #2f81f7;
    --color-tip: #3fb950;
    --color-warning: #d29922;
    --color-severe: #db6d28;
    --color-caution: #f85149;
    --color-important: #a371f7;
  }
}

</style>
<style>
.markdown-alert {
  padding: 0.5rem 1rem;
  margin-bottom: 16px;
  color: inherit;
  border-left: .25em solid #888;
}

.markdown-alert>:first-child {
  margin-top: 0
}

.markdown-alert>:last-child {
  margin-bottom: 0
}

.markdown-alert .markdown-alert-title {
  display: flex;
  font-weight: 500;
  align-items: center;
  line-height: 1
}

.markdown-alert .markdown-alert-title .octicon {
  margin-right: 0.5rem;
  display: inline-block;
  overflow: visible !important;
  vertical-align: text-bottom;
  fill: currentColor;
}

.markdown-alert.markdown-alert-note {
  border-left-color: var(--color-note);
}

.markdown-alert.markdown-alert-note .markdown-alert-title {
  color: var(--color-note);
}

.markdown-alert.markdown-alert-important {
  border-left-color: var(--color-important);
}

.markdown-alert.markdown-alert-important .markdown-alert-title {
  color: var(--color-important);
}

.markdown-alert.markdown-alert-warning {
  border-left-color: var(--color-warning);
}

.markdown-alert.markdown-alert-warning .markdown-alert-title {
  color: var(--color-warning);
}

.markdown-alert.markdown-alert-tip {
  border-left-color: var(--color-tip);
}

.markdown-alert.markdown-alert-tip .markdown-alert-title {
  color: var(--color-tip);
}

.markdown-alert.markdown-alert-caution {
  border-left-color: var(--color-caution);
}

.markdown-alert.markdown-alert-caution .markdown-alert-title {
  color: var(--color-caution);
}

</style>
        
        </head>
        <body class="vscode-body vscode-light">
            <!-- title: Identifying Spatially Variable Genes  -->
<h1 id="identifying-spatially-variable-genes-using-big-small-patch-bsp-algorithm-and-sparse-matrix-implementation-as-scbsp">Identifying Spatially Variable Genes using Big-Small Patch (BSP) algorithm and sparse matrix implementation as scBSP</h1>
<h2 id="bsp-algorithm">BSP Algorithm</h2>
<p>Big-small patch (BSP) is a granularity-guided, data-driven, and parameter-free model for identifying spatial variable genes in 2D and 3D high-throughput spatial transcriptomics data.</p>
<p><img src="https://github.com/juexinwang/BSP/blob/main/flowchart.png" alt="BSP"></p>
<p>Original implementation of BSP in python is available at <a href="https://github.com/juexinwang/BSP">https://github.com/juexinwang/BSP</a></p>
<h2 id="scbsp---a-fast-tool-for-single-cell-spatially-variable-genes-identifications-on-large-scale-spatially-resolved-transcriptomics-data">scBSP - A Fast Tool for Single-Cell Spatially Variable Genes Identifications on Large-Scale Spatially Resolved Transcriptomics Data</h2>
<p>This package utilizes a granularity-based dimension-agnostic tool, single-cell big-small patch (scBSP), implementing <strong>sparse matrix</strong> operation and KD-tree/balltree method for distance calculation, for the identification of spatially variable genes on
large-scale data. A corresponding Python library is available at <a href="https://pypi.org/project/scbsp/">https://pypi.org/project/scbsp</a>. The R source code is available at <a href="https://github.com/CastleLi/scBSP">https://github.com/CastleLi/scBSP</a>
. The Python source code is available at <a href="https://github.com/YQ-Wang/scBSP">https://github.com/YQ-Wang/scBSP</a></p>
<h1 id="system-requirement">System Requirement</h1>
<p>Tested on MacOS Sonoma version 14.4.1 with R version 4.3.1, XXX</p>
<h1 id="installation">Installation</h1>
<p>This package can be installed on R CRAN</p>
<pre><code>install.packages(&quot;scBSP&quot;)
</code></pre>
<h1 id="example-1-breast-cancer-data">Example 1: Breast Cancer Data</h1>
<p>Load the scBSP package and Breast cancer data set, which can be downloaded <a href="https://github.com/juexinwang/Tutorial_DahShu2024/blob/master/data/Layer2_BC_Count.rds">here</a>.</p>
<pre><code>library('scBSP')
load(&quot;./Layer2_BC_Count.rds&quot;)
</code></pre>
<p>View the expression count matrix rawcount, each row denotes a gene and each column represents a cell/spot.</p>
<pre><code>rawcount[1:5,1:5]

    17.907x4.967   18.965x5.003   18.954x5.995    17.846x5.993 20.016x6.019
GAPDH   1   7   5   1   2
USP4    1   0   0   0   0
MAPKAPK2    1   1   0   0   1
CPEB1   0   0   0   0   0
LANCL2  0   0   0   0   0
</code></pre>
<h2 id="prepare-the-input-coordinates-and-gene-expression">Prepare the input: coordinates and gene expression</h2>
<p>Extract the coordinates from the rawdata</p>
<pre><code>info &lt;- cbind.data.frame(x=as.numeric(sapply(strsplit(colnames(rawcount),split=&quot;x&quot;),&quot;[&quot;,1)),
                         y=as.numeric(sapply(strsplit(colnames(rawcount),split=&quot;x&quot;),&quot;[&quot;,2)))
rownames(info) &lt;- colnames(rawcount)
Coords=info[,1:2]
</code></pre>
<p>View the coordinates of x and y on 2-D space</p>
<pre><code>head(Coords)

                  x     y
17.907x4.967 17.907 4.967
18.965x5.003 18.965 5.003
18.954x5.995 18.954 5.995
17.846x5.993 17.846 5.993
20.016x6.019 20.016 6.019
20.889x6.956 20.889 6.956
</code></pre>
<p>View the dimension of coordinates</p>
<pre><code>dim(Coords)
[1] 251   2
</code></pre>
<h2 id="preprocessing">Preprocessing</h2>
<p>Excluding low expressed genes</p>
<pre><code>Filtered_ExpMat &lt;- SpFilter(rawcount)
</code></pre>
<p>(Optional) For this data, we need to explicitly convert the data format to Matrix</p>
<pre><code>Filtered_ExpMat &lt;- Matrix::Matrix(Filtered_ExpMat, sparse = TRUE)
</code></pre>
<p>Check the dimension of the input data</p>
<pre><code>dim(Filtered_ExpMat)
[1] 11769   251
</code></pre>
<h2 id="computing-p-values">Computing p-values</h2>
<p>The inputs are the coordinates and the expresson matrix, scBSP calculates the p-values</p>
<pre><code>P_values &lt;- scBSP(Coords, Filtered_ExpMat)
</code></pre>
<h2 id="output-the-final-results">Output the final results</h2>
<pre><code>head(P_values)
  GeneNames     P_values
1     GAPDH 2.704573e-09
2      USP4 2.452562e-01
3  MAPKAPK2 4.177737e-03
4     CPEB1 7.656481e-01
5    LANCL2 7.272278e-01
6      MCL1 9.308317e-06
</code></pre>
<p>Final results are sorted and filtered if P_values&lt;0.05</p>
<pre><code>results &lt;- P_values[P_values$P_values&lt;0.05,]
results &lt;- results[order(results$P_values),]
head(results)
     GeneNames     P_values
326     SPINT2 0.000000e+00
261      POSTN 6.439294e-15
2013    COL1A1 6.550316e-15
1347       B2M 7.660539e-15
2370       FN1 1.521006e-14
1072    EEF1A1 1.356693e-13

dim(results)
[1] 1765    2
</code></pre>
<h1 id="example-2-hdst-data-of-mouse-hippocampus">Example 2: HDST data of mouse hippocampus</h1>
<p>HDST provides a subcellular resolution spatial trianscriptomics data, which contains 181,367 spots and 13,243 genes with density XXX. The data can be downloaded at <a href="https://github.com/juexinwang/Tutorial_DahShu2024/blob/master/data/CN24_D1_unmodgtf_filtered_red_ut_HDST_final_clean.rds">here</a></p>
<pre><code>library('scBSP')
load(&quot;./CN24_D1_unmodgtf_filtered_red_ut_HDST_final_clean.rds&quot;)
</code></pre>
<p>View the expression count matrix as a sparse matrix, each row denotes a gene and each column represents a cell/spot.</p>
<pre><code>sp_count[1:5,1:5]

5 x 5 sparse Matrix of class &quot;dgCMatrix&quot;
        1000x100 1000x103 1000x113 1000x114 1000x116
Rcn2           .        .        .        .        .
Mycbp2         .        .        .        .        .
mt-Rnr2        .        .        .        .        .
Mprip          .        .        .        .        .
Mroh1          .        .        .        .        .
</code></pre>
<h2 id="prepare-the-input-coordinates-and-gene-expression-1">Prepare the input: coordinates and gene expression</h2>
<p>Extract the coordinates</p>
<pre><code>info &lt;- cbind.data.frame(x=as.numeric(sapply(strsplit(colnames(sp_count),split=&quot;x&quot;),&quot;[&quot;,1)),
                         y=as.numeric(sapply(strsplit(colnames(sp_count),split=&quot;x&quot;),&quot;[&quot;,2)))
rownames(info)  &lt;- colnames(sp_count)
Coords        &lt;- as.matrix(info)
</code></pre>
<p>View the coordinates of x and y on 2-D space</p>
<pre><code>head(Coords)
            x   y
1000x100 1000 100
1000x103 1000 103
1000x113 1000 113
1000x114 1000 114
1000x116 1000 116
1000x142 1000 142
</code></pre>
<p>View the dimension of coordinates</p>
<pre><code>dim(Coords)
[1] 181367      2
</code></pre>
<h2 id="preprocessing-1">Preprocessing</h2>
<p>Removing mitochondrial genes</p>
<pre><code>mt_idx      &lt;- grep(&quot;mt-&quot;,rownames(sp_count))
if(length(mt_idx)!=0){
    sp_count    &lt;- sp_count[-mt_idx,]
}
</code></pre>
<p>Excluding low expressed genes</p>
<pre><code>Filtered_ExpMat &lt;- SpFilter(sp_count)
</code></pre>
<p>Check the dimension of the input data</p>
<pre><code>dim(Filtered_ExpMat)
[1]  13209 181367
</code></pre>
<h2 id="computing-p-values-1">Computing p-values</h2>
<p>Calculate Spatially Variable Genes with this sparse matrix using scBSP.</p>
<pre><code>P_values &lt;- scBSP(Coords, Filtered_ExpMat)
</code></pre>
<h2 id="recording-computation-time">Recording computation time</h2>
<p>Install peakRAM package to record computational resources</p>
<pre><code>install.packages(&quot;peakRAM&quot;)
</code></pre>
<p>or</p>
<pre><code>install.packages(&quot;remotes&quot;)
remotes::install_github(&quot;tpq/peakRAM&quot;)
</code></pre>
<p>Record computational resources, it takes seconds on a laptop.</p>
<pre><code>library(peakRAM)
peakRAM(P_values &lt;- scBSP(Coords, Filtered_ExpMat))

Normalizing the expression matrix
Normalizing the coords matrix
Calculating p-values
                            Function_Call Elapsed_Time_sec Total_RAM_Used_MiB Peak_RAM_Used_MiB
1 P_values&lt;-scBSP(Coords,Filtered_ExpMat)            2.877                0.1         505559907
</code></pre>
<h2 id="output-the-final-results-1">Output the final results</h2>
<p>Final results are sorted and filtered if P_values&lt;0.05</p>
<pre><code>results &lt;- P_values[P_values$P_values&lt;0.05,]
results &lt;- results[order(results$P_values),]
head(results)
    GeneNames     P_values
24    Gm42418 0.000000e+00
70      Cmss1 0.000000e+00
74     Camk1d 3.330669e-16
49       Gphn 7.771561e-16
64 CT010467.1 1.975087e-13
43       Cdk8 1.445843e-12

dim(results)
[1] 1829    2
</code></pre>
<h1 id="example-3-slideseq-v2-data-on-mouse-olfactory-bulb">Example 3: SlideSeq V2 data on Mouse Olfactory Bulb</h1>
<p>Use SlideSeq V2 data with higher density.</p>
<h2 id="install-data-using-seuratdata">Install data using SeuratData</h2>
<pre><code>install.packages(&quot;remotes&quot;)
remotes::install_github(&quot;satijalab/seurat-data&quot;)

library(Seurat)
library(SeuratData)
library(SeuratObject)
library(peakRAM)
</code></pre>
<p>Check the available data</p>
<pre><code>AvailableData()
</code></pre>
<p>Install Slide-seq v2 dataset of mouse hippocampus</p>
<pre><code>InstallData(&quot;ssHippo&quot;)
</code></pre>
<h2 id="load-data-and-preprocessing">load data and preprocessing</h2>
<pre><code>slide.seq &lt;- LoadData(&quot;ssHippo&quot;)
data_extracted &lt;- scBSP::LoadSpatial(slide.seq)
ExpMatrix_Filtered &lt;- scBSP::SpFilter(data_extracted$ExpMatrix, Threshold = 1)
</code></pre>
<p>Check the dimensions</p>
<pre><code>dim(data_extracted$Coords)
[1] 53173     2

dim(ExpMatrix_Filtered)
[1] 23243 53173
</code></pre>
<p>Record the computation time</p>
<pre><code>peakRAM({ P_values &lt;- scBSP::scBSP(data_extracted$Coords,ExpMatrix_Filtered)})

Normalizing the expression matrix
Normalizing the coords matrix
Calculating p-values
                                                       Function_Call Elapsed_Time_sec Total_RAM_Used_MiB Peak_RAM_Used_MiB
1 {P_values&lt;-scBSP::scBSP(data_extracted$Coords,ExpMatrix_Filtered)}           15.572                  0         456867078
</code></pre>
<p>We can see it takes more time than HDST data for they have different density.</p>
<h1 id="cite">Cite</h1>
<p>Wang, J., Li, J., Kramer, S.T. et al. Dimension-agnostic and granularity-based spatially variable gene identification using BSP. Nat Commun 14, 7367 (2023). <a href="https://doi.org/10.1038/s41467-023-43256-5">https://doi.org/10.1038/s41467-023-43256-5</a></p>
<h1 id="references">References:</h1>
<ol>
<li><a href="https://github.com/juexinwang/BSP/">https://github.com/juexinwang/BSP/</a></li>
<li><a href="https://github.com/CastleLi/scBSP/">https://github.com/CastleLi/scBSP/</a></li>
<li>Stahl, P. L. et al. Visualization and analysis of gene expression in tissue sections by spatial transcriptomics. Science 353, 78-82, 2016</li>
<li><a href="https://github.com/mssanjavickovic/3dst">https://github.com/mssanjavickovic/3dst</a></li>
<li><a href="https://xzhoulab.github.io/SPARK/02_SPARK_Example/">https://xzhoulab.github.io/SPARK/02_SPARK_Example/</a></li>
</ol>

            <script async src="https://cdn.jsdelivr.net/npm/katex-copytex@latest/dist/katex-copytex.min.js"></script>
            
        </body>
        </html>